gulpfile.ts

```diff
@@ -0,0 +1,103 @@
+/// <reference path="./declarations.d.ts"/>
+/// <reference types="es6-promise" />
+
+import {execSync} from "child_process";
+import del = require("del");
+import * as fse from "fs-extra";
+import * as gulp from "gulp";
+import {createServer} from "http-server";
+//const httpServer = require("http-server");
+import * as path from "path";
+import * as tmp from "tmp";
+import ts = require("gulp-typescript");
+
+const out = "public";
+function outDir(name: string): string { return path.join(out, name); }
+
+gulp.task("clean", () => del(`${out}/*`));
+
+gulp.task("script", ["clean"], () => {
+    const tsProject = ts.createProject("assets/script/tsconfig.json", {
+        typescript: require("typescript")
+    });
+    return tsProject.src().pipe(ts(tsProject)).js.pipe(gulp.dest(outDir("script")));
+});
+
+function copy(src: string, dest: string): NodeJS.ReadWriteStream {
+    return gulp.src(src).pipe(gulp.dest(dest));
+}
+
+gulp.task("static", ["clean"], () => copy("assets/static/**", out));
+gulp.task("lib", ["clean"], () => copy("node_modules/@(jquery|typeahead.js)/**", outDir("lib")));
+gulp.task("index", ["clean"], () => copy("search-index-@(full|head|min).json", out));
+
+gulp.task("build", ["clean", "script", "static", "lib", "index"]);
+
+gulp.task("serve", () => {
+    const server = createServer({ root: "public" });
+    console.log("\nServing to localhost\n");
+    server.listen(80);
+});
+
+gulp.task("watch", ["build", "serve"], () => gulp.watch("assets/**", ["build"]));
+
+gulp.task("default", ["watch"]);
+
+gulp.task("publish", ["build"], () => {
+    function exec(cmd: string): string {
+        return execSync(cmd, { encoding: "utf8" });
+    }
+
+    if (!(exec("git status").includes("nothing to commit"))) {
+        throw new Error("Commit all changes first!")
+    }
+
+    if (exec("git rev-parse --abbrev-ref HEAD").trim() !== "master") {
+        throw new Error("You are not on master branch.");
+    }
+
+    const tmpObj = tmp.dirSync();
+    console.log(`Temporaries are stored at ${tmpObj.name}`);
+    function tmpDir(dir: string): string {
+        return path.join(tmpObj.name, dir);
+    }
+
+    const toMove = ["node_modules", "public"];
+    // Move files away temporarily.
+    const moved = Promise.all(toMove.map(dir => mvPromise(dir, tmpDir(dir))));
+
+    moved.then(() => {
```

Comment:
Return this promise, so `gulp` has something to look at to know when the task is done.